cmd/remote_val_harness/internal/test_harness.go

```diff
@@ -0,0 +1,363 @@
+		}
+		if err := th.TestPublicKey(); err != nil {
+			th.Shutdown(err)
+		}
+		if err := th.TestSignProposal(); err != nil {
+			th.Shutdown(err)
+		}
+		if err := th.TestSignVote(); err != nil {
+			th.Shutdown(err)
+		}
+		th.logger.Info("SUCCESS! All tests passed.")
+	}()
+
+	c := make(chan os.Signal, 1)
+	signal.Notify(c, os.Interrupt)
+	go func() {
+		for sig := range c {
+			th.logger.Info("Caught interrupt, terminating...", "sig", sig)
+			th.Shutdown(newTestHarnessError(ErrInterrupted, nil, ""))
+		}
+	}()
+
+	// Run until complete
+	<-donec
+	th.Shutdown(nil)
+}
+
+func (th *TestHarness) TestPublicKey() error {
+	th.logger.Info("TEST: Public key of remote signer")
+	th.logger.Info("Local", "pubKey", th.fpv.GetPubKey())
+	th.logger.Info("Remote", "pubKey", th.sc.GetPubKey())
+	if th.fpv.GetPubKey() != th.sc.GetPubKey() {
+		th.logger.Error("FAILED: Local and remote public keys do not match")
+		return newTestHarnessError(ErrTestPublicKeyFailed, nil, "")
+	}
+	return nil
+}
+
+func (th *TestHarness) TestSignProposal() error {
+	th.logger.Info("TEST: Signing of proposals")
+	// sha256 hash of "hash"
+	hash, _ := hex.DecodeString("D04B98F48E8F8BCC15C6AE5AC050801CD6DCFD428FB5F9E65C4E16E7807340FA")
+	prop := &types.Proposal{
+		Type:     types.ProposalType,
+		Height:   12345,
+		Round:    23456,
+		POLRound: -1,
+		BlockID: types.BlockID{
+			Hash: hash,
+			PartsHeader: types.PartSetHeader{
+				Hash:  hash,
+				Total: 1000000,
+			},
+		},
+		Timestamp: time.Now(),
+	}
+	propBytes, err := cdc.MarshalBinaryLengthPrefixed(types.CanonicalizeProposal(th.chainID, prop))
+	if err != nil {
+		th.logger.Error("FAILED: Could not marshal proposal to bytes", "err", err)
+		return newTestHarnessError(ErrTestSignProposalFailed, err, "")
+	}
+	if err := th.sc.SignProposal(th.chainID, prop); err != nil {
+		th.logger.Error("FAILED: Signing of proposal", "err", err)
+		return newTestHarnessError(ErrTestSignProposalFailed, err, "")
+	}
+	th.logger.Debug("Signed proposal", "prop", prop)
+	// first check that it's a basically valid proposal
+	if err := prop.ValidateBasic(); err != nil {
+		th.logger.Error("FAILED: Signed proposal is invalid", "err", err)
+		return newTestHarnessError(ErrTestSignProposalFailed, err, "")
+	}
+	// now validate the signature on the proposal
+	if th.sc.GetPubKey().VerifyBytes(propBytes, prop.Signature) {
+		th.logger.Info("Successfully validated proposal signature")
+	} else {
+		th.logger.Error("FAILED: Proposal signature validation failed")
+		return newTestHarnessError(ErrTestSignProposalFailed, err, "signature validation failed")
+	}
+	return nil
+}
+
+func (th *TestHarness) TestSignVote() error {
+	th.logger.Info("TEST: Signing of votes")
+	for _, voteType := range voteTypes {
+		th.logger.Info("Testing vote type", "type", voteType)
+		// sha256 hash of "hash"
+		hash, _ := hex.DecodeString("D04B98F48E8F8BCC15C6AE5AC050801CD6DCFD428FB5F9E65C4E16E7807340FA")
```

Comment:
This and the comment above can be replaced by:
tmhash.SumTruncated([]byte("hash"))

Reply:
Awesome! I'll definitely do that 👍